R/event_manager.R

```diff
@@ -90,6 +125,42 @@ EventManager <- R6Class(
     # notifications, others do not. (Not really sure why.)
     event_enable_domains = NULL,
 
+    event_domain_manually_enabled = list(),
+
+    maybe_event_domain_enable = function(domain) {
+      # If we're doing auto events
+      if (!private$session$get_auto_events()) return()
+      # and this domain requires events (not all domains require or have an .enable method)
+      if (!isTRUE(private$event_enable_domains[[domain]])) return()
+      # and events were not already manually enabled
+      if (isTRUE(private$event_domain_manually_enabled[[domain]])) return()
+      # and we're going from 0 to 1 callbacks in this domain
+      # or to 1+ the number of callbacks when `$disable()` was last called
+      threshold <- max(c(1, private$event_callback_count_threshold[[domain]]))
+      if (private$event_callback_counts[[domain]] != threshold) return()
+
+      # ...then automatically enable events.
+      private$session$debug_log("Enabling events for ", domain)
+      private$session[[domain]]$enable()
+      # We enabled this domain, not the user, so this stays FALSE
+      private$event_domain_manually_enabled[[domain]] <- FALSE
+    },
+
+    maybe_event_domain_disable = function(domain) {
+      # If we're doing auto events
+      if (!private$session$get_auto_events()) return()
+      # and the domain requires events
+      if (!isTRUE(private$event_enable_domains[[domain]])) return()
+      # and events were not manually enabled for the domain
+      if (isTRUE(private$event_domain_manually_enabled[[domain]])) return()
+      # and we're going from 1 to 0
+      if (private$event_callback_counts[[domain]] > 0) return()
```

Comment:
Unlike the section above, we're still not calling `$disable()` until the callback count is back down to 0, which may never happen if the user intervened with a `$disable()` call and we lost events. This seems like an acceptable edge case, and is better than not re-activating events.

I'm assuming that we can't set this to the callback count threshold, i.e. the callback count at the time when `$disable()` is called, because we may receive some events that were attached to the previous time period when events were enabled.